app/src/test/kotlin/compiler/parser/analysis/Common.kt

```diff
@@ -0,0 +1,58 @@
+package compiler.parser.analysis
+
+import compiler.common.dfa.DfaWalk
+import compiler.common.dfa.state_dfa.Dfa
+import compiler.common.dfa.state_dfa.DfaState
+import compiler.parser.grammar.Production
+
+typealias R = Production<GrammarSymbol>
+typealias GrammarSymbol = String
+const val EPSILON = ""
+
+class GrammarAnalysisTest {
+
+    class TestDfa(
+        override val startState: TestDfaState,
+        private val dfaDescription: Map<Pair<TestDfaState, GrammarSymbol>, TestDfaState>
+    ) : Dfa<GrammarSymbol, R> {
+
+        override fun newWalk(): DfaWalk<GrammarSymbol, R> = TestDfaWalk(startState, dfaDescription)
+    }
+
+    class TestDfaWalk(
+        private var currentState: TestDfaState,
+        private val dfaDescription: Map<Pair<TestDfaState, GrammarSymbol>, TestDfaState>
+    ) : DfaWalk<GrammarSymbol, R> {
+
+        override fun getAcceptingStateTypeOrNull(): R? = currentState.result
+
+        override fun isDead(): Boolean = !dfaDescription.keys.map { it.first }.contains(currentState)
+
+        override fun step(a: GrammarSymbol) {
+            dfaDescription[Pair(currentState, a)]?.let { currentState = it }
+        }
+    }
+
+    class TestDfaState(name: String) : DfaState<GrammarSymbol, R> {
+
+        override val result = if (name.startsWith("acc")) R() else null
+        override val possibleSteps: Map<GrammarSymbol, TestDfaState> = mapOf()
```

Comment:
All three of `nullable`, `first`, `follow` are heavily relying on this function, so it must be implemented correctly here.

Reply:
Implemented!